### ISSUE TITLE
Question: Since all default is JSONWebKey, its important to know the type of the original did key?

### ISSUE THREAD

**Issue by vongohren** [transmute-industries/verifiable-data#126]
When generating a key it is default type JSONWebKey2020, meaning that the type cannot declare what type this key should be converted to?

Either one have to use the did:key identifier in itself? Or stor on the side the intended type of the key?

Because a JWK for BLSG2 key cannot be created using from of a ED key right?

Just wanting to make sure I understand the logic 100% here

**Comment by OR13** [comment]
https://github.com/transmute-industries/verifiable-data/blob/main/packages/json-web-signature/src/__tests__/generate.test.ts

You should be able to  call `.export({ type: 'Ed25519VerifificationKey2018' })` on JsonWebKey's of that type.

**Comment by vongohren** [comment]
Yeah I know about the export. But you create a key, then export to another if need be right? Makes sense.

Just trying to understand what you guys do when you have created a key, and it has JSON structure and you store it in a KMS. It then belongs to a did, but the type itself is lost right? Because the JWK itself does not tell you what type of key was orignially created. If you store the JSONWebKey, and then pull it out again at some point, it would be difficult to know if it was BBS+ key or ED key. So one can store as ed25519 verificaiton key by exporting, but what would you suggest to be best practice @OR13 ?

**Comment by OR13** [comment]
> Because the JWK itself does not tell you what type of key was originally created.

correct, but you can assume that IF it can be a JWK, it can be a `JsonWebKey2020`.... The other "linked data key types"... are not safe bets, there is no way to know if they wanted to store that key as "Ed25519VerificationKey2018"... and that key type is not a standard, nor is `publicKeyBase58`... This is very frustrating.

I would recommend NEVER using a key format other that JWK.

But if you have to, you can rely on the did document to handle that, for did methods that support other key formats...

For example:

`did:example:123#key-0` -> `Ed25519VerificationKey2018`(publicKeyBase58) -> KMS -> (publicKeyJwk, privateKeyJwk)

`did:example:123#key-0` -> `JsonWebKey2020`(publicKeyJwk) -> KMS -> (publicKeyJwk, privateKeyJwk)

Not all did methods support all key types... which is another reason to pick a standard key type like JWK over non standard ones...`JsonWebKey2020` is non standard.... but `publicKeyJwk` is a standard.... using them together is the best we can do right now IMO.

In other words... The original key type is not relevant IF it can be represented as a JWK AND the DID Method that it is registered to preserves the public key and original type.

**Comment by vongohren** [comment]
Interessting, thanks for this information! I will take it into consideration.
Currently it seems that storing JWK for us, we need to label it to be able to use/initiate the right suite.
But what you are saying is that, because we have the DID, we can basically deduct that from the resolved DID, IF that methods is preserving the public key and original type?

**Comment by OR13** [comment]
I suggest looking into did:jwk, its like "did:key", but less dependencies and easer to use... see https://github.com/transmute-industries/did-transmute
